### PR TITLE
Migration from Jupyterlab v3 to v4

### PR DESCRIPTION
**Pull request by AjayThorve**
**_JupyterLab Extension Migration from v3 to v4 with Improved Backend and Enhanced Front-End Features_**

The existing extension, which exclusively functions with JupyterLab v3, has undergone a transformation to ensure compatibility with JupyterLab v4 and above. This migration not only includes updating the version but also incorporates various improvements in both the backend and frontend components of the extension

**Backend Migration:**
The previous version of our extension relied on a Bokeh server in the backend and utilized iframes to display GPU utilization charts. In this PR, we've transitioned to utilizing the built-in Tornado routes provided by JupyterLab. Additionally, we've integrated backend REST API calls with @tornado.web.authenticated, eliminating the need for a separate server and port management. This streamlined backend architecture simplifies deployment and maintenance.

**Frontend Enhancements:**
The frontend of the extension has been significantly enhanced. We've leveraged React Native widgets and the Recharts charting library to create a visually pleasing and smooth user experience. The use of React Native widgets ensures a more aesthetic design, eliminates frame drops during widget drag-and-drop operations, and takes advantage of React's virtual DOM for efficient rendering.

**New Features:**

1. **Brush for Time Series Charts**: We've introduced a brushing feature for time series charts, allowing users to easily inspect past events by selecting a specific time range.
2. **Synced Tooltips**: For pages with multiple charts, our extension now provides synchronized tooltips for timestamps across all charts. This feature enhances the user's ability to analyze data cohesively. 
3. **Theme Compatibility**: The extension now seamlessly integrates with JupyterLab themes. It adapts its colors and aesthetics based on whether the user is in a light or dark theme, ensuring a consistent and visually appealing experience.

**Version Compatibility:**
It's important to note that this updated extension is designed exclusively for JupyterLab v4 and later versions. To ensure continued support for JupyterLab v3 users, we will maintain the previous version separately.

**Future Improvements:**
We understand that there was a request to add cell execution markers to the charts. However, due to the complexities associated with asynchronous cells, we have decided to defer this feature to a future update. Rest assured, we will explore this enhancement in subsequent releases

Attaching some snapshots of the updated design, feel free to suggest design and aesthitic changes.

image
image
image
recording_2023-09-27_at_17 40 40

cc @exactlyallan @jacobtomlinson

### CODE REVIEW

**Comment by jakirkham** [comment]
Is this still planned for 0.10 or is it moving to 0.11?

**Comment by AjayThorve** [comment]
> Is this still planned for 0.10 or is it moving to 0.11?

Not sure, @jacobtomlinson ?

**Review by jacobtomlinson — COMMENTED** [review]
This is generally looking awesome! Thank you so much for putting in the effort here.

One thing that would be really great to see is a completely rewritten README with an updated demo gif, but more importantly updated contributing/developing guidelines. I really want to get to a better place to enable other folks in RAPIDS to contribute to this, so an explanation of how to set up a dev environment (bearing in mind TS and web dev is unfamiliar to many folks) and where to make changes is going to be critical to the maintainability of this work.

**Comment by jacobtomlinson** [comment]
> Is this still planned for 0.10 or is it moving to 0.11?

We are on `0.9` now so I think `0.10` should be fine.

**Comment by AjayThorve** [comment]
> This is generally looking awesome! Thank you so much for putting in the effort here.
> 
> One thing that would be really great to see is a completely rewritten README with an updated demo gif, but more importantly updated contributing/developing guidelines. I really want to get to a better place to enable other folks in RAPIDS to contribute to this, so an explanation of how to set up a dev environment (bearing in mind TS and web dev is unfamiliar to many folks) and where to make changes is going to be critical to the maintainability of this work.

That's great feedback, I'll get to working on a new README and a DEVELOPER_GUIDELINES markdowns

**Comment by jakirkham** [comment]
RAPIDS has completed code freeze with a release planned soon. Hence the question of what release is being targeted

**Comment by jacobtomlinson** [comment]
@jakirkham this project is released ad-hoc and generally isn't tired to the bi-monthly RAPIDS releases.

**Comment by jakirkham** [comment]
Ok wasn't sure

**Review by jacobtomlinson — APPROVED** [review]
This looks great, thanks for working on the README and CONTRIBUTING docs.

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      image: rapidsai/ci-conda:latest
+      env:
+        RAPIDS_BUILD_TYPE: ${{ inputs.build_type || 'branch' }}
+        PARALLEL_LEVEL: ${{ env.PARALLEL_LEVEL }}
+    steps:
+      - uses: aws-actions/configure-aws-credentials@v3
+        with:
+          role-to-assume: ${{ vars.AWS_ROLE_ARN }}
+          aws-region: ${{ vars.AWS_REGION }}
+          role-duration-seconds: 43200 # 12h
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:60]: Let's decrease this value to something reasonable for the job duration.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      - uses: actions/checkout@v3
+        with:
+          ref: ${{ inputs.sha }}
+          fetch-depth: 0
+      - name: Standardize repository information
+        run: |
+          echo "RAPIDS_REPOSITORY=${{ github.repository }}" >> "${GITHUB_ENV}"
+          echo "RAPIDS_SHA=$(git rev-parse HEAD)" >> "${GITHUB_ENV}"
+          echo "RAPIDS_REF_NAME=${{ inputs.branch || github.ref_name }}" >> "${GITHUB_ENV}"
+          echo "RAPIDS_NIGHTLY_DATE=${{ inputs.date }}" >> "${GITHUB_ENV}"
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:70]: There is a shared action for this:

- https://github.com/rapidsai/shared-actions/tree/main/rapids-github-info
- usage example - https://github.com/rapidsai/shared-workflows/blob/6b6fa90493e3457aa84cd8532705dcda4b14cc9e/.github/workflows/wheels-publish.yaml#L64

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      env:
+        RAPIDS_BUILD_TYPE: ${{ inputs.build_type || 'branch' }}
+        PARALLEL_LEVEL: ${{ env.PARALLEL_LEVEL }}
+    steps:
+      - uses: aws-actions/configure-aws-credentials@v3
+        with:
+          role-to-assume: ${{ vars.AWS_ROLE_ARN }}
+          aws-region: ${{ vars.AWS_REGION }}
+          role-duration-seconds: 43200 # 12h
+      - uses: actions/checkout@v3
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:97]: These shared actions be updated to their latest versions.

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -0,0 +1,120 @@
+name: build
+
+on:
+  push:
+    branches:
+      - "branch-*"
+    tags:
+      - v[0-9][0-9].[0-9][0-9].[0-9][0-9]
```
[the remark this answers is not in the document]
Comment (ajschmidt8) [reply .github/workflows/build.yaml:8]: Link to docs: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-and-excluding-paths

[the remark this answers is not in the document]
Comment (AjayThorve) [reply .github/workflows/build.yaml]: `v[0-9]+.[0-9]+.[0-9]+a` added this, allows one or more digits

File: .github/workflows/build.yaml [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  issues: none
+  packages: read
+  pages: none
+  pull-requests: read
+  repository-projects: none
+  security-events: none
+  statuses: none
+
+jobs:
+  python-build:
```
[the remark this answers is not in the document]
Comment (AjayThorve) [reply .github/workflows/build.yaml:29]: added conda-build and wheel-build resusable workflows

File: .github/workflows/build.yaml (hunk 3 above)
Comment (AjayThorve) [reply .github/workflows/build.yaml]: Updated to v4

File: .github/workflows/build.yaml (hunk 2 above)
Comment (AjayThorve) [reply .github/workflows/build.yaml]: thanks, using it now

File: .github/workflows/build.yaml (hunk 1 above)
Comment (AjayThorve) [reply .github/workflows/build.yaml]: decreased to 20 minutes

File: .github/workflows/pr.yaml [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    needs: wheel-build
+    runs-on: 'linux-amd64-gpu-v100-latest-1'
+    container:
+      image: rapidsai/ci-wheel:latest
+      env:
+        RAPIDS_BUILD_TYPE: pull-request
+        PARALLEL_LEVEL: ${{ env.PARALLEL_LEVEL }}
+        NVIDIA_VISIBLE_DEVICES: ${{ env.NVIDIA_VISIBLE_DEVICES }}
+    steps:
+      - uses: aws-actions/configure-aws-credentials@v3
```
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:95]:
```suggestion
      - uses: aws-actions/configure-aws-credentials@v4
```

File: .github/workflows/pr.yaml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    needs: conda-python-build
+    runs-on: 'linux-amd64-gpu-v100-latest-1'
+    container:
+      image: rapidsai/ci-conda:latest
+      env:
+        RAPIDS_BUILD_TYPE: pull-request
+        PARALLEL_LEVEL: ${{ env.PARALLEL_LEVEL }}
+        NVIDIA_VISIBLE_DEVICES: ${{ env.NVIDIA_VISIBLE_DEVICES }}
+    steps:
+      - uses: aws-actions/configure-aws-credentials@v3
```
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:60]:
```suggestion
      - uses: aws-actions/configure-aws-credentials@v4
```

File: .github/workflows/build.yaml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/wheels-build.yaml@branch-24.02
+    with:
+      build_type: ${{ inputs.build_type || 'branch' }}
+      branch: ${{ inputs.branch }}
+      sha: ${{ inputs.sha }}
+      date: ${{ inputs.date }}
+      script: ci/build_wheel.sh
+      # Package is pure Python and only ever requires one build.
+      matrix_filter: 'map(select((.LINUX_VER | test("centos")|not))) | sort_by((.PY_VER | split(".") | map(tonumber))) | .[-1] + {ARCH: "amd64"}'
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:57]:
```suggestion
      matrix_filter: 'map(select((.LINUX_VER | test("centos")|not))) | sort_by((.PY_VER | split(".") | map(tonumber))) | [.[-1] + {ARCH: "amd64"}']
```

This should fix the latest CI issue. The result of this filter needs to be an array, not an object.

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/wheels-build.yaml@branch-24.02
+    with:
+      build_type: ${{ inputs.build_type || 'branch' }}
+      branch: ${{ inputs.branch }}
+      sha: ${{ inputs.sha }}
+      date: ${{ inputs.date }}
+      script: ci/build_wheel.sh
+      # Package is pure Python and only ever requires one build.
+      matrix_filter: 'map(select((.LINUX_VER | test("centos")|not))) | sort_by((.PY_VER | split(".") | map(tonumber))) | [.[-1] + {ARCH: "amd64"}']
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:57]:
```suggestion
      matrix_filter: 'map(select((.LINUX_VER | test("centos")|not))) | sort_by((.PY_VER | split(".") | map(tonumber))) | [.[-1] + {ARCH: "amd64"}]'
```

Whoops. one more fix.

File: .github/workflows/pr.yaml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    with:
+      build_type: pull-request
+  wheel-build:
+    needs: checks
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/wheels-build.yaml@branch-24.02
+    with:
+      build_type: pull-request
+      # Package is pure Python and only ever requires one build.
+      matrix_filter: 'map(select((.LINUX_VER | test("centos")|not))) | sort_by((.PY_VER | split(".") | map(tonumber))) | .[-1] + {ARCH: "amd64"}'
```
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:44]:
```suggestion
      matrix_filter: 'map(select((.LINUX_VER | test("centos")|not))) | sort_by((.PY_VER | split(".") | map(tonumber))) | [.[-1] + {ARCH: "amd64"}]'
```

Forgot this instance.

**Review by jacobtomlinson — APPROVED** [review]

**Comment by XQ-UT** [comment]
Can we do a release that contains this fix? Does the current v0.11.0a include this fix?

**Comment by AjayThorve** [comment]
> Can we do a release that contains this fix? Does the current v0.11.0a include this fix?

@XQ-UT You can access this via the nightly packages already:

```bash
# nightly version (for jupyterlab>=4)
pip install --extra-index-url https://pypi.anaconda.org/rapidsai-wheels-nightly/simple --pre jupyterlab_nvdashboard
```

There are a couple of issues, once fixed we will do a stable release in a week or so.

**Pull request merged**